lib/cf_deployer/driver/auto_scaling_group.rb

```diff
@@ -62,21 +62,42 @@ def desired_capacity_reached?
         healthy_instance_count >= desired_capacity
       end
 
-      private
+      def healthy_instance_ids
+        instances = auto_scaling_instances.select do |instance|
+          'HEALTHY'.casecmp(instance.health_status) == 0
+        end
+        instances.map(&:id)
+      end
+
+      def in_service_instance_ids
+        elbs = load_balancers
+        return [] if elbs.empty?
+
+        ids = elbs.collect(&:instances)
+                  .collect(&:health)
+                  .to_a
+                  .collect { |elb_healths|
+                     elb_healths.select { |health| health[:state] == 'InService' }
+                                .map { |health| health[:instance].id }
+                  }
+
+        ids.inject(:&)
+      end
 
       def healthy_instance_count
-        begin
-          count = auto_scaling_instances.count do |instance|
-            instance.health_status == 'HEALTHY' && (load_balancers.empty? || instance_in_service?( instance.ec2_instance ))
-          end
-          Log.info "Healthy instance count: #{count}"
-          count
-        rescue => e
-          Log.info "Unable to determine healthy instance count due to error: #{e.message}"
-          -1
+        AWS.memoize do
+          instances = healthy_instance_ids
+          instances &= in_service_instance_ids unless load_balancers.empty?
```

Comment:
ok, per our hipchat conversation, it's clear now that the `AutoScalingGroupSwap` strategy (given how it's currently written) can involve an overlap period where 2 ASGs from different stacks are in the same ELB, so the complicated logic of intersecting the ELB and ASG healthchecks makes more sense now.